### PR TITLE
Telemetry integration to the UI

### LINKED ISSUE #14900: Add telemetry for generic usage metrics collections in Graylog
Add the following features to core to support telemetry:

- Saving user preference for telemetry data metrics
- Add tools to configure and send metric
- Instrument metric collection for events. (ex. pageview)

### PR DESCRIPTION
**Pull request by ousmaneo**
This PR implements anonymous usage data collection on the UI part. 

fix #14900 
/jenkins-pr-deps Graylog2/graylog-plugin-enterprise#4943
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog2/telemetry/rest/TelemetryResponseFactory.java
Comment (kroepke) [inline graylog2-server/src/main/java/org/graylog2/telemetry/rest/TelemetryResponseFactory.java:94]: I think this does not need the `%s`, `CLUSTER` nesting, we will only ever have a single cluster here. If we ever have multi-cluster setups that users interact with, that will need a different representation anyay.
Remove the cluster id here makes the reported properties much simpler, too.

File: graylog2-web-interface/src/logic/telemetry/TelemetryProvider.tsx
Comment (kroepke) [inline graylog2-web-interface/src/logic/telemetry/TelemetryProvider.tsx:74]: could we rename this to `'cluster'` before merging, please?

File: graylog2-server/src/main/java/org/graylog2/telemetry/rest/TelemetryResource.java
Comment (moesterheld) [inline graylog2-server/src/main/java/org/graylog2/telemetry/rest/TelemetryResource.java:100]: This is the endpoint which is also called when editing another user's settings, so either this needs to be changed to accept a user id or disabled in the UI.
Right now, changing the value for any user in the UI will result in changing my own setting.

Comment (moesterheld) [reply graylog2-server/src/main/java/org/graylog2/telemetry/rest/TelemetryResource.java:100]: Oh yeah, same for `getTelemetrySettings`

File: graylog2-web-interface/src/components/indices/IndexSetsComponent.jsx
Comment (grotlue) [inline graylog2-web-interface/src/components/indices/IndexSetsComponent.jsx:70]: I think we should not make tracking dependent on any calls or actions. If I understand correctly, we want to track user behavior, so I'd track on click immediately, as the user clicks anyways, no matter if calls fail or not.

File: graylog2-web-interface/src/components/inputs/CreateInputControl.jsx
Comment (grotlue) [inline graylog2-web-interface/src/components/inputs/CreateInputControl.jsx:106]: Same here, the tracking should not depend on the action.

File: graylog2-web-interface/src/components/inputs/InputStateControl.jsx
Comment (grotlue) [inline graylog2-web-interface/src/components/inputs/InputStateControl.jsx:74]: Same here, we should track the click as soon as the user clicks.

File: graylog2-web-interface/src/components/inputs/InputStateControl.jsx
Comment (grotlue) [inline graylog2-web-interface/src/components/inputs/InputStateControl.jsx:88]: Dependent tracking.

File: graylog2-web-interface/src/components/outputs/OutputsComponent.jsx
Comment (grotlue) [inline graylog2-web-interface/src/components/outputs/OutputsComponent.jsx:109]: Dependent tracking.

File: graylog2-web-interface/src/components/outputs/OutputsComponent.jsx
Comment (grotlue) [inline graylog2-web-interface/src/components/outputs/OutputsComponent.jsx:124]: Dependent tracking.

File: graylog2-web-interface/src/components/roles/RoleEdit/UsersSection.tsx
Comment (grotlue) [inline graylog2-web-interface/src/components/roles/RoleEdit/UsersSection.tsx:64]: Dependent tracking.

File: graylog2-web-interface/src/logic/telemetry/TelemetryInfoText.tsx
Comment (grotlue) [inline graylog2-web-interface/src/logic/telemetry/TelemetryInfoText.tsx:28]: I have some suggestions for improvements in the copy:

```
We would like to collect anonymous usage data to help us prioritize improvements and make Graylog better in the future.

We do not collect personal data, sensitive information, or content such as logs in your instances.

Learn more on our [Privacy Policy](https://www.graylog.com/).

You can turn data collection off or on at any time in the user profile.
```

File: graylog2-web-interface/src/logic/telemetry/TelemetryInfoText.tsx
Comment (grotlue) [inline graylog2-web-interface/src/logic/telemetry/TelemetryInfoText.tsx:35]: We need to update the link to link to a proper Privacy Policy.

File: graylog2-web-interface/src/logic/telemetry/TelemetryInit.test.tsx
Comment (grotlue) [inline graylog2-web-interface/src/logic/telemetry/TelemetryInit.test.tsx:74]: Can we make this expect more explicit?

File: graylog2-web-interface/src/pages/IndexSetCreationPage.tsx
Comment (grotlue) [inline graylog2-web-interface/src/pages/IndexSetCreationPage.tsx:66]: Dependent tracking.

File: graylog2-web-interface/src/threatintel/components/ThreatIntelPluginConfig.jsx
Comment (grotlue) [inline graylog2-web-interface/src/threatintel/components/ThreatIntelPluginConfig.jsx:99]: Dependent tracking.

File: graylog2-web-interface/src/util/StringUtils.ts
Comment (grotlue) [inline graylog2-web-interface/src/util/StringUtils.ts:63]: Why not use `map` instead?

File: graylog2-web-interface/src/views/components/CreateNewWidgetModal.tsx
Comment (grotlue) [inline graylog2-web-interface/src/views/components/CreateNewWidgetModal.tsx:86]: Can we just not set the attribute at all?

**Review by grotlue — CHANGES_REQUESTED** [review]
The general approach looks good! 
Code-wise the main point I found is that some of our tracking depends on API calls, which I think we shouldn't do, as we are tracking clicks and e.g. not only successful submits. I tried to comment all lines in the code, but not sure if I found them all.

I tested the pageview tracking, info modal setting and the tracking on search page. I have not tested all the other sub pages as I understood that the tracking is not complete. If there are specific pages I should have a look at, let me know which :)

# Testing Results 

## Info modal
- ✅ Telemetry info modal is shown on first load
- ❌ “Privacy Policy” link still links to [graylog.org](http://graylog.org/)
- ✅ The modal does not appear again after closing it (when logging out and back in)
- ✅ The modal is shown to each user independently

## Tracking
- ✅Telemetry can be turned off in the user profile (except for the admin user)
- ❌ Tracking is still sent after turning telemetry off in the user profile
- ✅ Does not track anything when logged out (on login page)
- ✅ Identifies the user when logging in 
- ✅ Page view event is tracked on page load

### Modals and popovers
Tested those on search page

- ❌ When closing a modal or popover we have no information if it was closed via cancel or save/confirm. The event_info only contains a constructer, no other values.

### Search Page
- ❌ When resetting the search stream filter it is tracked as “change_input_value”. The event_info only has a constructor, and no info about the value that was selected. Having this will enable us to differentiate.
- ⚠️ No tracking inside of the “Search Time Range” form
- ⚠️ No tracking for the search button
- ⚠️ No tracking for the widgets
- ⚠️ No tracking for search filters

File: graylog2-server/src/main/java/org/graylog2/telemetry/rest/TelemetryResource.java
Comment (AntonEbel) [reply graylog2-server/src/main/java/org/graylog2/telemetry/rest/TelemetryResource.java:100]: Good catch. I spoke to Ousmane, we are adjusting the UI so that a user can only change his own telemetry settings.

Comment (ousmaneo) [reply graylog2-server/src/main/java/org/graylog2/telemetry/rest/TelemetryResource.java:100]: Hi @moesterheld, now we are not allowing a user to change the setting for other. You can only change your own if your profile is not read-only.

**Review by grotlue — CHANGES_REQUESTED** [review]
✅ Tracking is now not fired when I turn it off in user profile

I have one small text suggestions.

Besides that I think we should make sure we distinguish between modal/popover cancel & "submit" actions for closing as they can tell a lot about user behaviour.

The rest looks good! Thanks a lot!

File: graylog2-server/src/test/java/org/graylog2/telemetry/rest/TelemetryServiceTest.java
Comment (moesterheld) [inline graylog2-server/src/test/java/org/graylog2/telemetry/rest/TelemetryServiceTest.java:105]: Why do telemetry data have to be collected when the permission has not been asked?

**Review by moesterheld — APPROVED** [review]
two small comments, otherwise backend lgtm

File: graylog2-server/src/main/java/org/graylog2/telemetry/rest/TelemetryResponseFactory.java
Comment (AntonEbel) [reply graylog2-server/src/main/java/org/graylog2/telemetry/rest/TelemetryResponseFactory.java]: As discussed, the cluster node data is no longer flattened.

File: graylog2-web-interface/src/util/StringUtils.ts
Comment (gally47) [reply graylog2-web-interface/src/util/StringUtils.ts:63]: Because the output array isn't guaranteed to have the same length as the input array.

**Comment by grotlue** [comment]
> > ✅ Tracking is now not fired when I turn it off in user profile
> > I have one small text suggestion.
> > Besides that I think we should make sure we distinguish between modal/popover cancel & "submit" actions for closing as they can tell a lot about user behaviour.
> > The rest looks good! Thanks a lot!
> 
> @grotlue I am not sure if we should distinguish between modal/popover cancel & submit actions for closing, but since it can't be done globally in the modal component, I suggest we add a new PR if it's needed.

I think we should definitely distinguish between those actions. If it's too much effort for this PR we can do this in another one, but I'd really think that we create an issue for this and add this soon. We have a lot of modals and popovers and knowing when a user cancels an action (e.g. closing modal via cancel button) will give us a lot of insights.

File: graylog2-web-interface/src/views/components/searchbar/SearchButton.tsx
Comment (grotlue) [inline graylog2-web-interface/src/views/components/searchbar/SearchButton.tsx:63]: We can call `sendTelemetry` here directly, no need for an extra function.

File: graylog2-web-interface/src/views/components/widgets/WidgetActionsMenu.tsx
Comment (grotlue) [inline graylog2-web-interface/src/views/components/widgets/WidgetActionsMenu.tsx:192]: Is this the correct eventElement?

**Review by grotlue — APPROVED** [review]
Frontend part looks good.

I have two small comments, I think the 2nd one should be looked at, but after that it's ready to go.

As said I think we need an issue for improving the modal tracking.

Besides that, I think it would be great if we have a documentation about our tracking: What events we track, what data do we send with which event, how to name elements. This way we can make sure that other engineers can add tracking properly to new features in the future.

Thanks a lot for working on this topic! Excited to see it go live!

File: graylog2-web-interface/src/views/components/searchbar/SearchButton.tsx
Comment (ousmaneo) [reply graylog2-web-interface/src/views/components/searchbar/SearchButton.tsx:63]: Actually, we can't because the function is defined outside of the component so we need to pass the function or move it inside the scope of the component.

**Comment by ousmaneo** [comment]
> Besides that, I think it would be great if we have a documentation about our tracking: What events we track, what data do we send with which event, how to name elements. This way we can make sure that other engineers can add tracking properly to new features in the future.


Thank you for your review, I will prepare the documentation for the telemetry.

File: graylog2-server/src/main/java/org/graylog2/telemetry/rest/TelemetryService.java
Comment (AntonEbel) [reply graylog2-server/src/main/java/org/graylog2/telemetry/rest/TelemetryService.java]: fixed

File: graylog2-server/src/test/java/org/graylog2/telemetry/rest/TelemetryServiceTest.java
Comment (AntonEbel) [reply graylog2-server/src/test/java/org/graylog2/telemetry/rest/TelemetryServiceTest.java:105]: We do not check permission here, as a user can only check his own telemetry settings.

File: graylog2-web-interface/src/logic/telemetry/TelemetryProvider.tsx
Comment (dennisoelkers) [inline graylog2-web-interface/src/logic/telemetry/TelemetryProvider.tsx:66]: Can we wrap this call with a `try {} catch { console.warn(...); }`? It is now used all over the product and we should make sure that adding telemetry does not interfere with other functionality.

Comment (dennisoelkers) [inline graylog2-web-interface/src/logic/telemetry/TelemetryProvider.tsx:66]: What happens when the user has telemetry disabled? We are skipping the group configuration in the `useEffect` block in that case, but should we also just return a no-op function here instead of calling `posthog.capture()`?

**Review by dennisoelkers — CHANGES_REQUESTED** [review]
The code looks good in general. Due to the size I was not able to look at the complete code and focussed on the core parts in general, adding two major questions/suggestions in there.

File: graylog2-web-interface/src/logic/telemetry/TelemetryProvider.tsx
Comment (ousmaneo) [reply graylog2-web-interface/src/logic/telemetry/TelemetryProvider.tsx:66]: Thank you, @dennisoelkers, for having a look and for the suggestions.

> What happens when the user has telemetry disabled? We are skipping the group configuration in the useEffect block in that case, but should we also just return a no-op function here instead of calling posthog.capture()?


When telemetry is disabled, the init is not called, and the PostHog context should be empty, resulting in `usePostog` returning `posthog` with the value of `undefined`, therefor calling `posthog.capture` is not called.

File: graylog2-server/src/main/java/org/graylog2/telemetry/rest/TelemetryResource.java
Comment (ousmaneo) [reply graylog2-server/src/main/java/org/graylog2/telemetry/rest/TelemetryResource.java:100]: Thank you @moesterheld . I fixed the view part

**Review by dennisoelkers — APPROVED** [review]
Looks good to me!

**Pull request merged**
